parsl/dataflow/dflow.py

```diff
@@ -714,6 +714,22 @@ def atexit_cleanup(self):
         if not self.cleanup_called:
             self.cleanup()
 
+    def cleanup_wait_for_all_tasks(self):
+        """Waits for all tasks in the task list to be completed, by waiting for their
+        AppFuture to be completed. This method will not necessarily wait for any tasks
+        added after cleanup has started (such as data stageout?)
+        """
+
+        logger.info("Waiting for all remaining tasks to complete")
+        for task_id in self.tasks:
+            # .exception() is a less exception throwing way of
+            # waiting for completion than .result()
+            fut = self.tasks[task_id]['exec_fu']
+            if not fut.done():
+                logger.debug("Waiting for task {} to complete".format(task_id))
+                fut.exception()
```

Comment:
If a task raises an exception, will that be logged anywhere else? If not, maybe we should do so here.

Reply:
We shouldn't switch to an eager fail behavior here, unless the user requested it. We should log the exception and wait for the next one.
```
try:
     fut.result()
except Exception as e:
     logger.exception("Task raised this :{}".format(e))
```

Reply:
It's not an eager fail as implemented in the PR at the moment - fut.exception() doesn't throw an exception even if the future has an exception rather than a real result. (I think)

Reply:
I think, if we agree it makes sense to log an exception, we should go with Yadu's suggestion to use `result()` to avoid confusion about what `.exception()` does

Reply:
we don't really have a way to detect "lost" tasks at the moment (whatever that means? tasks that the workflow has never waited on?)

we can get a list of every task ever, and a list of every task that hasn't completed at DFK shutdown, but neither of those is a particularly good candidate for outputting exceptions from - anything that failed before that won't have an exception logged here.

So maybe we should rely on the existing logging (or not logging) of exceptions that happen if you don't ask for a result, so that it is consistent before/after shutdown?

Reply:
Ben correctly points out that logging the error here doesn't make sense, it will be a random subset of tasks depending on whether it is done or not.

Reply:
yep, agreed